lib/src/main/java/graphql/nadel/engine/transform/partition/NadelPartitionTransform.kt

```diff
@@ -0,0 +1,266 @@
+            pathToPartitionArg = getPathToPartitionArg(overallField, executionBlueprint.engineSchema) ?: return null,
+            fieldPartitionContext = fieldPartitionContext,
+            graphQLSchema = executionBlueprint.engineSchema,
+            partitionKeyExtractor = partitionTransformHook.getPartitionKeyExtractor()
+        )
+
+        val fieldPartitions = try {
+            fieldPartition.createFieldPartitions(field = overallField)
+        } catch (exception: Exception) {
+            return State(
+                executionContext = executionContext,
+                fieldPartitionContext = fieldPartitionContext,
+                errors = mutableListOf(exception)
+            )
+        }
+
+        if (fieldPartitions.size < 2) {
+            // We can't partition a field that doesn't have at least two partitions
+            return null
+        }
+
+        partitionTransformHook.willPartitionCallback(executionContext, fieldPartitions)
+
+        return State(
+            executionContext = executionContext,
+            fieldPartitionContext = fieldPartitionContext,
+            fieldPartitions = fieldPartitions
+        )
+    }
+
+    private fun getPathToPartitionArg(overallField: ExecutableNormalizedField, schema: GraphQLSchema): List<String>?  {
+        val fieldDef = overallField.getFieldDefinitions(schema).first()
+        val routingDirective = fieldDef.getAppliedDirective(NadelDirectives.partitionDirectiveDefinition.name) ?: return null
+        val pathToSplitPoint = routingDirective.getArgument("pathToSplitPoint")
+            ?.argumentValue
+            ?.value as ArrayValue?
+
+        return pathToSplitPoint?.values?.map { it as StringValue }?.map { it.value }
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        serviceExecutionContext: NadelServiceExecutionContext,
+        transformer: NadelQueryTransformer,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+
+        if (state.errors != null) {
+            // At this point we know we can't partition the field, but removing it could result
+            // in a query with no top-level fields, which would be invalid.
+            // Adding __typename as an artificial field also wouldn't work because that field
+            // belongs to all operation types, which causes issues further down the execution.
+            return NadelTransformFieldResult.unmodified(field)
+        }
+
+        val fieldPartitions = checkNotNull(state.fieldPartitions) { "Expected fieldPartitions to be set" }
+
+        val primaryPartition = fieldPartitions.values.first()
+
+        // TODO: throw error if operation is Subscription?
+        val rootType = executionContext.query.operation.getType(executionBlueprint.engineSchema)
+
+        val partitionCalls = coroutineScope {
```

Comment:
As mentioned easier on call, this `coroutineScope` is only going to return once the child `async` jobs are done.

So we need to launch jobs off a `supervisorJob` earlier.

Like maybe

```diff
diff --git a/lib/src/main/java/graphql/nadel/NextgenEngine.kt b/lib/src/main/java/graphql/nadel/NextgenEngine.kt
index f57a6a5b..54f28ae2 100644
--- a/lib/src/main/java/graphql/nadel/NextgenEngine.kt
+++ b/lib/src/main/java/graphql/nadel/NextgenEngine.kt
@@ -70,6 +70,7 @@ import kotlinx.coroutines.future.await
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.reactive.asFlow
 import kotlinx.coroutines.reactive.asPublisher
+import kotlinx.coroutines.supervisorScope
 import java.util.concurrent.CompletableFuture
 import graphql.normalized.ExecutableNormalizedOperationFactory.Options.defaultOptions as executableNormalizedOperationFactoryOptions
 
@@ -144,7 +145,7 @@ internal class NextgenEngine(
         queryDocument: Document,
         instrumentationState: InstrumentationState?,
         executionHints: NadelExecutionHints,
-    ): ExecutionResult {
+    ): ExecutionResult = supervisorScope {
         try {
             val timer = NadelInstrumentationTimer(
                 instrumentation,
@@ -177,6 +178,7 @@ internal class NextgenEngine(
                 timer,
                 incrementalResultSupport,
                 resultTracker,
+                executionCoroutine = this,
             )
 
             val beginExecuteContext = instrumentation.beginExecute(
@@ -226,7 +228,7 @@ internal class NextgenEngine(
             // todo: maybe pass in the incremental version that's built below into here
             resultTracker.complete(result)
 
-            return if (incrementalResultSupport.hasDeferredResults()) {
+            if (incrementalResultSupport.hasDeferredResults()) {
                 IncrementalExecutionResultImpl.Builder()
                     .from(result)
                     .incrementalItemPublisher(incrementalResultSupport.resultFlow().asPublisher())
@@ -236,7 +238,7 @@ internal class NextgenEngine(
             }
         } catch (e: Throwable) {
             when (e) {
-                is GraphQLError -> return newExecutionResult(error = e)
+                is GraphQLError -> newExecutionResult(error = e)
                 else -> throw e
             }
         }
diff --git a/lib/src/main/java/graphql/nadel/engine/NadelExecutionContext.kt b/lib/src/main/java/graphql/nadel/engine/NadelExecutionContext.kt
index a9c26e76..4e6e0bc8 100644
--- a/lib/src/main/java/graphql/nadel/engine/NadelExecutionContext.kt
+++ b/lib/src/main/java/graphql/nadel/engine/NadelExecutionContext.kt
@@ -11,6 +11,7 @@ import graphql.nadel.hooks.CreateServiceContextParams
 import graphql.nadel.hooks.NadelExecutionHooks
 import graphql.nadel.result.NadelResultTracker
 import graphql.normalized.ExecutableNormalizedOperation
+import kotlinx.coroutines.CoroutineScope
 import java.util.concurrent.CompletableFuture
 import java.util.concurrent.ConcurrentHashMap
 
@@ -24,7 +25,8 @@ data class NadelExecutionContext internal constructor(
     internal val incrementalResultSupport: NadelIncrementalResultSupport,
     internal val resultTracker: NadelResultTracker,
     internal val hydrationDetails: ServiceExecutionHydrationDetails? = null,
-    internal val isPartitionedCall: Boolean = false
+    internal val isPartitionedCall: Boolean = false,
+    internal val executionCoroutine: CoroutineScope,
 ) {
     private val serviceContexts = ConcurrentHashMap<String, CompletableFuture<Any?>>()
```

Then we can do thing like

```kotlin
val partitionCalls = otherPartitions.map {
    executionContext.executionCoroutine.async {
        val topLevelField = NFUtil.createField(
            executionBlueprint.engineSchema,
            rootType,
            // TODO: queryPath contains field aliases, not field names, which results in an error.
            field.queryPath,
            it.normalizedArguments,
            it.children
        )

        engine.executePartitionedCall(topLevelField, service, state.executionContext)
    }
}
```